guardian/gateway
# Refactor sign in to use Okta session cookie check

## What does this change?

This PR changes the way we check for existing sessions when a user visits the signin page.

If a user already has an Okta session, and tries to sign in with another account, Okta will refresh the old session, rather than adding the new one. For this reason we want to check for an existing session before we try and sign the user in.

We previously did not have access to the `sid` Okta session cookie as it was set on `auth.theguardian.com` domain name, and gateway is on the `profile.theguardian.com` domain. Now we have switch the domain of the cookie to be `profile.theguardian.com` we can read the session id from the cookie, and check if the session is valid via the Okta sessions API [`/api/v1/sessions/:sessionId`](https://developer.okta.com/docs/reference/api/sessions/#get-session) via the `getSession` method added in https://github.com/guardian/gateway/pull/1536

The changes in the PR:

- Adds the session check via the `getSession` call.
- Keeps the functionality to redirects the user to the `returnUrl` parameter if they have an existing session. Note that `redirectUrl` is parsed so only allowed values can be used!
- Moves the sign in mock tests to their own file
- Remove the redirect from `/signin` that sent the user through the authorization code flow to check their session.
- Removes the `signInRedirect` parameter from the `encryptedStateCookie`.
- Updates the documentation to reflect the new flow.                  

## How to test

- Sign in on `profile.theguardian.com/signin?useOkta=true`
- After successful sign in revist `profile.theguardian.com/signin?useOkta=true`
- You should be redirected to `theguardian.com/uk` and your current session should still exist
- You should not have been prompted to sign in
